iOS/MMKV/MMKV/MMKV.mm

```diff
@@ -89,66 +90,94 @@ + (instancetype)defaultMMKV {
 
 // any unique ID (com.tencent.xin.pay, etc)
 + (instancetype)mmkvWithID:(NSString *)mmapID {
-	if (mmapID.length <= 0) {
-		return nil;
-	}
-	CScopedLock lock(g_instanceLock);
-
-	MMKV *kv = [g_instanceDic objectForKey:mmapID];
-	if (kv == nil) {
-		kv = [[MMKV alloc] initWithMMapID:mmapID cryptKey:nil];
-		[g_instanceDic setObject:kv forKey:mmapID];
-	}
-	return kv;
+    return [self mmkvWithID:mmapID cryptKey:nil];
 }
 
 + (instancetype)mmkvWithID:(NSString *)mmapID cryptKey:(NSData *)cryptKey {
-	if (mmapID.length <= 0) {
-		return nil;
-	}
-	CScopedLock lock(g_instanceLock);
-
-	MMKV *kv = [g_instanceDic objectForKey:mmapID];
-	if (kv == nil) {
-		kv = [[MMKV alloc] initWithMMapID:mmapID cryptKey:cryptKey];
-		[g_instanceDic setObject:kv forKey:mmapID];
-	}
-	return kv;
-}
-
-- (instancetype)initWithMMapID:(NSString *)mmapID cryptKey:(NSData *)cryptKey {
-	if (self = [super init]) {
-		m_lock = [[NSRecursiveLock alloc] init];
-
-		m_mmapID = mmapID;
-
-		m_path = [MMKV mappedKVPathWithID:m_mmapID];
-		if (!isFileExist(m_path)) {
-			createFile(m_path);
-		}
-		m_crcPath = [MMKV crcPathWithMappedKVPath:m_path];
-
-		if (cryptKey.length > 0) {
-			m_cryptor = new AESCrypt((const unsigned char *) cryptKey.bytes, cryptKey.length);
-		}
-
-		[self loadFromFile];
-
+    return [self mmkvWithID:mmapID cryptKey:cryptKey relativePath:nil];
+}
+
++ (instancetype)mmkvWithID:(NSString *)mmapID relativePath:(nullable NSString *)path {
+    return [self mmkvWithID:mmapID cryptKey:nil relativePath:path];
+}
+
++ (instancetype)mmkvWithID:(NSString *)mmapID cryptKey:(NSData *)cryptKey relativePath:(nullable NSString *)path {
+    if (mmapID.length <= 0) {
+        return nil;
+    }
+    if (![self canCreateMMKVPath:mmapID relativePath:path]) {
+        return nil;
+    }
+    
+    CScopedLock lock(g_instanceLock);
+    
+    NSString *kvKey = [self mmapKeyWithMMapID:mmapID relativePath:path];
+    MMKV *kv = [g_instanceDic objectForKey:kvKey];
+    if (kv == nil) {
+        kv = [[MMKV alloc] initWithMMapID:mmapID cryptKey:cryptKey relativePath:path];
+        [g_instanceDic setObject:kv forKey:kvKey];
+    }
+    return kv;
+}
+
++ (BOOL)canCreateMMKVPath:(NSString *)mmapID relativePath:(nullable NSString *)relativePath {
+    if (relativePath) {
+        NSString *path = [MMKV mappedKVPathWithID:mmapID relativePath:relativePath];
+        if (isFileExist(path)) {
+            return true;
+        } else {
+            BOOL ret = createFile(path);
```

Comment:
Doing creating inside a function named `canXXX` is confusing. Replace `createFile` with getting write permission of the specific location is better.

Reply:
`initWithMMapID` will create the file, by the way.

Reply:
1. why use `createFile` inside `canXXX` ?
Because we should make sure that a custom path could be created in platform, and I don't find a better method to check a custom path current or not.  I have to use `createFile`.

2. why should make sure a custom path current ?
As i considered, once a custom path could not be created in platform, MMKV should not cache data. 
But when I test code, if we don't check custom path correct,  MMKV still set/get/change data in Dictionary in Memory.  This will mistake user everything is OK, but actually the data won't  be stored in Disk.

3. `initWithMMapID` still need create the file for default path.

Reply:
If you have a **write permission** of the dir, you can create file inside.

Reply:
You know what, forget about it. I'll merge it and fix it later.

Reply:
I got what you said. i consider one user mistake that relative path value is '/' in iOS, so i decided create file in init method for custom path.